### PR TITLE
Add Error Object 

### LINKED ISSUE #829: Need Fingerprint mismatch error
Error if fingerprint in SDP does not match IDP Token. Need to know which mid had the failed fingerprint.

### PR DESCRIPTION
**Pull request by fluffy**
Fixes #829 #828 #827 #823 #822 #555

### CODE REVIEW

**Comment by fluffy** [comment]
I'd like to get some comments on this and see if people think it is heading the right direction.  There  are still more errors to do but if we can get theses ones in the right style, the rest are about the same so I would like to make sure these are about right first.

**Comment by fluffy** [comment]
Really, I thought we had agreed that we needed to understand SDP errors, particularly when they are remote SDP to recover. What do the minutes of the meeting say ?

**Comment by fluffy** [comment]
Glad to back the SDP errors out of this PR until we sort it out but I think any of us that have deployed production systems are going to want a lot more detail than this about what goes wrong with SDP.

**Comment by alvestrand** [comment]
The request for cases with "more data" was to show an use case where an application (not a programmer debugging the issue) would take two different actions based on the value of the returned data.

**Comment by stefhak** [comment]
@burnburn to review. @adam-be to contact script-coord

File: webrtc.html [hunk 1]
```diff
@@ -8905,20 +8918,26 @@ <h4 id="sec.idp-loginneeded">User Login Procedure</h4>
...
+        the promise with an <code>RTCError</code> with errorDetail set to
+        "IdPNeedLogin" and the <code>idpLoginUrl</code> attribute set to the URL
+        where the user can log on.  This URL might link to page where a user can
+        enter their (IdP) username and password, or otherwise provide any
+        information the IdP needs to authorize a assertion request. </p>
+
+        <p>The value of the <code>idpLoginUrl</code> attribute contained in the
+        rejection will also be passed to the application in the
+        <code>idpLoginUrl</code> attribute of the
+        <code>RTCPeerConnection</code>.</p>
```
Comment (alvestrand) [inline webrtc.html]: Devil's advocate: If this URL is stable and not subject to race conditions, is there a reason to have the information both here and in the error?

File: webrtc.html [hunk 2]
```diff
@@ -783,6 +783,14 @@
                     then reject <var>p</var> with a
                     <code>InvalidStateError</code> and abort these steps.</p>
                   </li>
+
+                  <li> <p>If the content of <var>description</var> is not valid
+                  SDP syntax, then reject <var>p</var> with an
+                  <code>RTCError</code> with errorDetail set to "sdpSyntaxError"
+                  and the sdpLineNumber attribute set to the line number in the
+                  SDP where the syntax error was detected then abort these
+                  steps.</p> </li>
```
Comment (alvestrand) [inline webrtc.html]: I still don't see a case where machine action can be taken on the SDP line number. What would change if we instead said "it is RECOMMENDED that the SDP line number be part of the message part of the Error"?

**Comment by aboba** [comment]
@fluffy Do we have an updated PR for discussion at the November 9 Virtual Interim? Slides are here:
https://docs.google.com/presentation/d/1SSokCdnEEg8SO8laBWCWK0emU-W-4uuuqdOIIJG4bOY/edit#slide=id.p

File: webrtc.html (hunk 2 above)
Comment (fluffy) [inline webrtc.html]: Lets say some SDP had 4 m line sections, the first is audio, second is main video, and next two came from polycom 3 screen system with proprietary extensions that are not valid SDP. If we an error on a given line, we will often just remove the whole m-section that line was in and try again. We might have look at the line and remove it and try again.

**Comment by fluffy** [comment]
> On Nov 7, 2016, at 8:37 PM, aboba <notifications@github.com> wrote:
> 
> @fluffy Do we have an updated PR for discussion at the November 9 Virtual Interim? Slides are here:
> https://docs.google.com/presentation/d/1SSokCdnEEg8SO8laBWCWK0emU-W-4uuuqdOIIJG4bOY/edit#slide=id.p
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

No update, this is still the latest PR on this one. (But JSEP is in WGLC which is where my time has been :-)

File: webrtc.html (hunk 1 above)
Comment (fluffy) [inline webrtc.html]: On Nov 9 WG call, decided to remove the idpLoginUrl from this PR and move to a separate bug that can be considered on it's own ( And waits till we decide if we do multiple identity providers or not first)

File: webrtc.html (hunk 2 above)
Comment (fluffy) [inline webrtc.html]: Decided on nov 9 WG call to stay with this as is.

**Comment by fluffy** [comment]
@dontcallmedom - Dom could you take a 10 second look at the travis error and see if it is obvious to you what I need to do? Thanks

**Comment by dontcallmedom** [comment]
@fluffy Travis CI should be safely ignored for now

**Comment by alvestrand** [comment]
@fluffy there are some trailing blanks that confuse the diff. Can you tidy?

**Comment by aboba** [comment]
Conflicts resolved and changes commited.

File: webrtc.html [hunk 3]
```diff
@@ -10216,6 +10251,345 @@ <h2 id="sec.idp-error-handling">IdP Error Handling</h2>
...
+          "#error-handling">%RTCError%</a>.</p>
+        </section>
+        <section>
+          <h2>RTCError.prototype.errorDetail</h2>
+          <p>The initial value of the errorDetail property of the prototype for 
+          the RTCError constructor is the empty String.</p>
+        </section>
+        <section>
+          <h2>RTCError.prototype.sdpLineNumber</h2>
+          <p>The initial value of the errorDetail property of the prototype for
```
Comment (taylor-b) [inline webrtc.html]: s/errorDetail/sdpLineNumber

File: webrtc.html [hunk 4]
```diff
@@ -10216,6 +10251,345 @@ <h2 id="sec.idp-error-handling">IdP Error Handling</h2>
...
+          the RTCError constructor is the empty String.</p>
+        </section>
+        <section>
+          <h2>RTCError.prototype.sdpLineNumber</h2>
+          <p>The initial value of the errorDetail property of the prototype for 
+          the RTCError constructor is 0.</p>
+        </section>
+        <section>
+          <h2>RTCError.prototype.httpRequestStatusCode</h2>
+          <p>The initial value of the errorDetail property of the prototype for
```
Comment (taylor-b) [inline webrtc.html]: s/errorDetail/httpRequestStatusCode

File: webrtc.html (hunk 3 above)
Comment (aboba) [inline webrtc.html]: Fixed in PR https://github.com/w3c/webrtc-pc/pull/973

File: webrtc.html (hunk 4 above)
Comment (aboba) [inline webrtc.html]: Fixed in PR https://github.com/w3c/webrtc-pc/pull/973

**Comment by aboba** [comment]
@fluffy Can you take a look at the latest draft to determine whether the material in this PR has been properly integrated?
https://rawgit.com/w3c/webrtc-pc/master/webrtc.html

**Comment by fluffy** [comment]
@aboba - Yes, I think it looks good.

**Pull request closed without merging**